### DISCUSSION TITLE
Usage of XMLHttpRequest in a WebComponent with Google Polymer 1.x

### DISCUSSION THREAD

**Discussion by eitch** [OWASP/www-project-csrfguard#205] [Q&A] [answered]
This is an interesting project, thank you for your time building it.

One question that i have: What configuration is needed for XMLHttpRequest in a WebComponent context.

I have a single page application, which uses Google Polymer 1.x (https://polymer-library.polymer-project.org/1.0/docs/devguide/feature-overview). This has a WebComponent iron-ajax (https://github.com/PolymerElements/iron-ajax/tree/1.x), which uses XMLHttpRequest internally. 

I was able to configure my project to using "Scenario: everything is protected, except...".

In some cases it works, and then suddenly the request token is not updated, generating "Request Token does not match Page Token". If i use TokenPerPage=true, then my issues are gone, but i am wondering if this makes the solution less secure.

**Comment by forgedhallpass** [comment, marked as answer]
Hello @eitch,

Unfortunately I am not familiar with Google Polymer, but if it uses `XMLHttpRequest`s internally and the `csrfguard.js` script runs first, which [overrides some of the functions](https://github.com/OWASP/www-project-csrfguard/blob/1f3324f006de8edd7e908fcb72fa749dd9596a91/csrfguard/src/main/resources/csrfguard.js#LL102C36-L102C36) then it should theoretically work.

The documentation of the `TokenPerPage` configuration can be found in the [`csrfguard.properties`](https://github.com/OWASP/www-project-csrfguard/blob/1f3324f006de8edd7e908fcb72fa749dd9596a91/csrfguard/src/main/resources/csrfguard.properties#L95) file. If it is disabled then there will not be unique tokens assigned to different endpoints, and the solution will only use a single `masterToken`. Both can prevent basic CSRF attacks, but if you have an XSS vulnerability in your solution, it will be hardER (not impossible) to exploit if the `TokenPerPage` is enabled, because more logic is needed for it.
